### PR TITLE
`install-agent` UX improvements

### LINKED ISSUE #345: Failed installs report the installer type

### LINKED ISSUE #346: Before beginning installation, the development environment in which the installation will be performed is summarized and confirmed by the user
**Java**:
- `JAVA_HOME`
- `JDK Version`

**Python**:
- `python -c 'import sys; print(sys.prefix)'`
- `Python version`

**Ruby**:
- `gem env home`
- `Ruby version`

`AgentInstaller` should be responsible for owning this method as part of the interface.
```ts
getEnvironment(): Promise<Record<string, string>>
```

### LINKED ISSUE #350: `appmap.cli.log` is populated
This telemetry value is currently being reported blank.
https://github.com/applandinc/appmap-js/blob/main/packages/cli/src/cmds/agentInstaller/install-agent.ts#L186

### PR DESCRIPTION
**Pull request by dustinbyrne**
Resolves #345
Resolves #350
Resolves #346   

This PR fixes a number of telemetry issues and adds a confirmation step before executing install which describes the installation environment.
image

### CODE REVIEW

**Review by apotterri — CHANGES_REQUESTED** [review]
This is definitely a big step forward. Thanks!

A few other things:
* we should consider sending `process.env.PATH` and a listing of the files in the project directory in the telemetry
* if both `poetry.lock` and `requirements.txt` are present, we should default to `poetry` as the project type
* in the first step, we should say how we determined the project type:
```
ajp@Alans-MacBook-Pro shardingsphere % node ~/src/applandinc/worktrees/appmap-js/packages/cli/built/src/cli.js install-agent java
? Select the type of Java project to install the AppMap agent to. (Use arrow keys)
❯ Maven
  Gradle
```
becomes something like
```
ajp@Alans-MacBook-Pro shardingsphere % node ~/src/applandinc/worktrees/appmap-js/packages/cli/built/src/cli.js install-agent java
Found "pom.xml", so this looks like a Maven project.
? Select the type of Java project to install the AppMap agent to. (Use arrow keys)
❯ Maven
  Gradle
```

* in
```
? AppMap is about to be installed. Confirm the details below.
  Project type: poetry
  Project directory: /Users/ajp/src/mirumee/saleor
  Python version: 3.9.6
  Python path: /Users/ajp/src/mirumee/saleor/.direnv/python-3.9.6
```
"Python path" might be easier to understand if it was, e.g. "Python package directory". Also, for all project types, I think it would be good to show what command is about to be run to install the agent:
```
Installation command: poetry add --allow-prereleases --dev appmap
```
* I'm still really not a fan of the "rerun with -v" flow: https://asciinema.org/a/USVQ07PdkThQQqo2YnBJHQxLQ

**Comment by kgilpin** [comment]
When an error occurs, can we prompt the user to print the full error instead of making them do it all again with -v ?

**Comment by apotterri** [comment]
> When an error occurs, can we prompt the user to print the full error instead of making them do it all again with -v ?

As I said before, I'd also really like to see it work this way.

File: packages/cli/src/telemetry.ts [hunk 1]
```diff
@@ -202,13 +198,19 @@ export default class Telemetry {
...
           console.log(JSON.stringify(event, null, 2));
         }
 
+        if (this.enabled) {
+          Telemetry.client.trackEvent(event);
+          Telemetry.client.flush({ callback: () => resolve() });
+        } else {
+          // Flush is never called, so we need to resolve the promise. We'll add a timeout so that
+          // the rest of the function can complete before we call back.
+          setTimeout(resolve);
```
[the remark this answers is not in the document]
Comment (dustinbyrne) [reply packages/cli/src/telemetry.ts:211]: The use of `setTimeout` was to guarantee order of execution, in that the rest of the function would complete before `resolve` was called. However, in thinking about it, I also can't guarantee `flush` is going to call its callback asynchronously, so I guess it becomes pointless. I've simplified this.

File: packages/cli/src/cmds/agentInstaller/pythonAgentInstaller.ts [hunk 2]
```diff
@@ -13,9 +13,42 @@ const REGEX_PKG_DEPENDENCY = /^\s*appmap\s*[=>~]+=.*$/m;
...
   constructor(readonly path: string) {}
 
+  async getPythonOutput(...args: string[]): Promise<string | undefined> {
+    try {
+      const { stdout, stderr } = await run(
+        new CommandStruct('python', args, this.path)
+      );
+      return stdout ? stdout : stderr;
+    } catch {
+      return undefined;
```
[the remark this answers is not in the document]
Comment (dustinbyrne) [reply packages/cli/src/cmds/agentInstaller/pythonAgentInstaller.ts]: Two error states will propagate to the installation environment summary now:
- `${program} was not found`
- `${program} failed with code ${code}`

Further details will be available in the exec log.

**Comment by dustinbyrne** [comment]
I've made changes to most of the comments above. Anything unaddressed at this time has been added to an issue.

**Review by apotterri — APPROVED** [review]
Looks good, thanks. Don't forget to rebase to squash the `fixup` commits --  they're a nuisance to get rid of after your merge.

**Pull request merged**